README.md
# Switchboard -- Re-imagined Virtual Switches for Vera Home Automation

![Switchboard](https://www.toggledbits.com/assets/switchboard/switchboard-default.png) 
This plugin is my take on a modernized *Virtual Switch* plugin for Vera and openLuup. The long-lived (near zombie) Virtual Switch plugin
has some subtle issues that make it difficult to use with some third party applications (including Google Home and Amazon Echo devices), and even Vera's own native mobile app for Android.
Hopefully, this version of virtual switches will address those shortcomings.

## Installing on Vera

Vera users can install Switchboard by searching for it under *Apps > Install apps*.

Alternately, users may request the following URL in a browser, substituting your Vera's local IP where indicated:

```
http://vera-local-ip/port_3480/data_request?id=action&action=CreatePlugin&PluginNum=9194&serviceId=urn:micasaverde-com:serviceId:HomeAutomationGateway1
```

## Installing on openLuup

**NOTE:** Version 2018.11.21 of openLuup, or higher, is required to run Switchboard.

Installation of Switchboard is best done from the AltAppStore. Find it. Click it.

## Creating Virtual Switches/Devices

To create virtual switches, go into the "Switchboard" device control panel, select the type of virtual device you want to create (switch, dimmer, etc.), and hit the "Create New Virtual Device" button. Each click requires processing and a Luup reload, so do this slowly if you have a large number of switches to create.

**IMPORTANT**: After creating virtual devices, you *must* do a [hard-refresh of your browser](https://www.howtogeek.com/672607/how-to-hard-refresh-your-web-browser-to-bypass-your-cache/) so that the Vera UI picks up and displays the newly-created devices.

**NOTE:** Do not create additional instances of the Switchboard master device. The system should have one and only one Switchboard device. Virtual switches are created only from the Switchboard control panel buttons as described above.

## Switchboard Virtual Switch Features

Switchboard's virtual switches use entirely Vera-native device types and service definitions, and this is what makes them work better than the old Virtual Switch devices (which had a custom device type that Vera really didn't recognize everywhere).

Switchboard's virtual switches can be hidden using the "command and control" interface in the "Status" tab of the Switchboard control panel. Here you will see all of your virtual switches in one place, and be able to control them all, and set options.

Currently, there are only two options: visibility, and self-reset. The visibility option allows you to hide virtual switches, so if you have a large number of them, you can easily get them off your UI7 device list without having to tuck them into a virtual room, etc. The self-reset option creates an "pulse" switch--a switch that provides an "on" pulse and then automatically turns itself off. You can control the length of the pulse for each switch independently; a pulse length of zero means the switch doesn't use pulse timing.

NOTE: Turning "on" a pulse switch that is already on does not extend pulse timing.

NOTE: When a switch is hidden, it will also not be visible in Vera's scene trigger menus and other places in the UI, so if you're trying to create a new scene using a hidden virtual switch, you will first need to go into the Switchboard status panel and un-hide the switch. You can re-hide it after; that doesn't affect the scene's ability to *use* the switch.

## Virtual Window Coverings

It may seem odd to have virtual window coverings, but in Vera, the window covering implementation uses the `Dimming1` service to set the opening (e.g. 0% is closed, and 50% is half open, and 100% is fully open). Calling `SetLoadLevelTarget` as one would for a dimmer controls the shade opening. The `SwitchPower` action `SetTarget` can also be used to quickly fully-open or fully-close the covering.

By default, Switchboard will simulate motor movement of the covering by ramping `LoadLevelStatus` at a rate of 5% per second to the target value. If your application requires a different rate, it can be set by setting `RampRatePerSecond` (as percent per second). Setting it to 0 disables ramp and causes the covering to go immediately to the requested target value.

## Legacy Virtual Switch Features

To maintain compatibility with the older Virtual Switch plugin's switches, Switchboard's switches implement the services and behaviors of the older plugin, in addition to the standard Vera binary switch behaviors.

The old Virtual Switch plugin supported two text fields for each switch. These have made their way into Switchboard's switches as well, for compatibility. However, because the Vera-native UI is used for Switchboard's virtual switches, there is no way to display them in the native Vera switch UI. They are visible and editable in the Switchboard "Status" tab, however.

If you have existing virtual switches from the old "Virtual Switch" plugin, you can "adopt" those switches and make them Switchboard devices, which then allows you to uninstall the old "Virtual Switch" plugin. The device numbers will stay the same, so no changes to scenes, Lua, Reactor, PLEG, etc. should be needed.

## Virtual Scene Controllers

Switchboard allows you to create virtual scene controllers. This gives you a multi-button UI on a single device that allows you to trigger scenes or Reactor/PLEG rules, etc. I use this for establishing "modes" that are unrelated to house modes--sometimes you want/need more than the standard Vera house mode delivers. For example, I now use a virtual scene controller to set and track if my home theater is in "Setup", "Roll Previews", "Roll Show", "Pause" or "Off" mode and make changes when transitioning between. Over my whole house, I use a virtual scene controller to let my other Reactors know if it is currently "Morning", "Day", "Evening" or "Night".

The virtual scene controllers (VSCs) can have any number of buttons, although 24 seems to be a practical limit for the UI, so that's the UI limit (you can still create more, only the first 24 are shown on the dashboard card). You determine the number of buttons a VSC can have by providing a comma-separated list of button labels in the VSC's control panel UI. Changing the labels, unfortunately, requires a Luup reload, and you must do a hard-refresh of your browser, for proper display. 

When a VSC is in *single state* mode, only one button can be "down" at a time--only one "mode" can be active at a time. When a button is pressed, the VSC sets its `sl_SceneActivated` state variable to the index number of button pressed (starting from 1). If a previous state was in effect, `sl_SceneDeactivated` is also set to the previous index for that old mode. In this way, a VSC in single state mode operates very similar to many physical scene controllers in the Vera world. The UI analogue for this behavior is "radio buttons". Also see the description of the `Value` and `Active` state variables below.

When a VSC is in *multi-state* mode, the UI allows more than one mode to be set. Pressing a mode button in the UI toggles the mode's on/off state. It is thus possible for multiple modes, or no mode at all, to be active at any given time. The UI analogue for this is checkboxes. In multi-state mode, the VSC sets `sl_SceneDeactivated` for the mode when it is turned off, and `sl_SceneActivated` as each is turned on. These values both containly only the state of the most-recently changed mode.

The `Value` state variable is also used to store the effective mode. For single-state VSCs, this will only ever contain one value: the last mode selected. For multi-state VSCs, it will contain a comma-separated list of all modes that are active.

The VSC also sets the state variable `ActiveN`, where `N` is the index number of the mode, to 1 on each active mode, or 0 otherwise.

> Now the bad news. Despite using the Vera-defined device definition, Vera's own declared UI for scene controllers is empty both in its web UI and mobile apps &mdash; the standard UIs show only the device name and an icon. In order to make Switchboard's VSCs useful at least in the web UI, Switchboard replaces the Vera standard UI declaration on VSC devices with a custom, dynamically-created one. But, as usual for all custom UIs in Vera currently, the current mobile apps (including third party) do not use the available UI definition data and so do not (and likely never will) be able to paint the custom buttons. Sorry, there will be no UI for VSCs in the mobile app world.

### Controlling VSCs

VSCs can be controlled by Lua, Reactor Activities, PLEG, etc. by invoking the `SetLight` action (in service `urn:micasaverde-com:serviceId:SceneControllerLED1`). The action takes two parameters: `Indicator` and `newValue`. Their use is as follows:

`Indicator`|`newValue`|Description/Function
---------|--------|-----------
An integer from 1 to the number of buttons/labels/modes *or* a string matching one of the defined modes/labels|0 or 1|For multi-state VSCs, turns the mode in `Indicator` on or off according to `newValue` (1=active, 0=inactive, blank=toggle). For single-state VSCs, the mode in `Indicator` is made the (only) current mode, and `newValue` must be 1 (any other value is invalid/unsupported).
The string `"Set$Labels"`|label list|Configures the list of modes/labels for the target VSC. This will cause a Luup reload. You will also need to hard refresh your browser after this action (but I can't make that happen from the plugin).
The string `"Set$Mode"`|0 or 1|Set single-state mode (`newValue` = 0) or multi-state mode (`newValue` = 1).
The string `"Set$Mask"`|maskbits (integer)|Sets the enabled modes to those having one bits in the (integer) mask bits in `newValue` (1=LSB). For example, the value 5 (binary 101) would turn on modes 1 and 3, and turn off all other modes. This is for multi-state VSCs only. Its use is undefined/unsupported for single-state VSCs.
The string `"Inc$Mode"`|ignored|Set the next ordinal mode (the mode numbered after the current one). If the next mode is off the end of the list, it wraps around and sets the first mode. This action is defined/supported only for single-state VSCs.
The string `"Dec$Mode"`|ignored|Set the previous ordinal mode (the mode numbered before the current one). If the next mode is off the end of the list, it wraps around and sets the first mode. This action is defined/supported only for single-state VSCs.

> In case you're wondering, VSCs use the standard definition of a scene controller device, but the "standard" (non-LED) scene controller Vera device type defines no actions at all, and the one and only action supported by the extension "scene controller with LEDs" device type is `SetLight`. That is why we must use `SetLight` as the action--it's the only thing we can use. We can't add actions to the standard definitions, so the only option is to "overload" the meaning of values passed to `SetLight`.

## Distinguishing Switchboard Virtual Devices from "Real" Devices

Sometimes it may be necessary to distinguish Switchboard's virtual devices from real devices (e.g. in a startup Lua routine). Device type cannot be used, as Switchboard tries to use Vera-standard devices types to the greatest extent possible. But there are some built-in "tells":

1. The best, most-reliable method is to check the parent of the device; if the parent is a Switchboard master device, then it's a Switchboard virtual device. A quick version of this test is `if (luup.devices[luup.devices[devicenum_being_tested].device_num_parent] or {}).device_type == "urn:schemas-toggledbits-com:device:Switchboard:1" then --[[ it's a Switchboard device --]] end`
2. An alternate method of detection is to look for the existence of the `Behavior` state variable in service `urn:toggledbits-com:serviceId:Switchboard1` on the device; if present and not blank, it's a Switchboard virtual device;
2. The old, now deprecated method is to check the *manufacturer* attribute on the device, which will be "rigpapa"; the *model* attribute will be set to the switch type name, which is always a string beginning with "Switchboard".

## License and Warranty

Switchboard, Copyright 2018,2019,2020 Patrick H. Rigney (rigpapa). All Rights Reserved.

Switchboard is offered under the MIT License:
> Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
# Switchboard -- Change Log

**NOTE:** Version 1.9 and all later versions are not posted to the Vera App Marketplace. Please install from Github directly. OpenLuup users can still use the AltAppStore to install Switchboard.

## Version 1.9 (released, 22203)

* Make dimmer `ImpulseTime` work.
* For non-integer ramp rates, make sure `LoadLevelStatus` only stores (rounded) integer.
* Support for ALTUI 2.52
* Fix docs for creating new virtual devices.

## Version 1.8 (released, 20179)

* Timer availability is now per-device-type, and removed from UI where not applicable.
* Completely restructure inits at switch startup.

## Version 1.7 (released)

* Tri-state switches will now timer-reset to the `TimerResetState` when set to any state other than that (not just "on" as in 1.6).
* New virtual Scene Controller! Set the `Values` state variable to the (comma-separated) list of button values. You can do this directly, or through the device's own UI (not the Switchboard master device--go into the device control panel on the virtual scene controller). See the README for more.

## Version 1.6 (released)

* Add Lock, Water Valve, and Relay virtual devices. Note that lock has a separate service for `Target` and `Status` from all other binary on/off devices &mdash; it does not use `urn:upnp-org:serviceId:SwitchPower1`, but rather `urn:micasaverde-com:serviceId:DoorLock1`.
* Make post-timer return state configurable via `TimerResetState`; this enables TriState virtual switches to return to off or void.

## Version 1.5 (released)

* Fix an issue with window covering ramp rate not properly handling zero value (disable ramp).

## Version 1.4 (released)

* Add support for virtual window covering (Vera implements these as dimmers, so while it seems an odd thing to do here, it's a natural for Vera).

## Version 1.3 (released)

* Fix TriState initialization

## Version 1.2 (released)

* Support repeat triggering. By default, the virtual switch will only set its status if it changes, and as a result, scenes will only trigger on changes in status. In some cases, it is useful to trigger scenes/watches whenever the switch state is set, regardless of whether or not that results in a change in status (repeat triggering), so add a new flag for this purpose.
* Fix notices and links in copyright area.

## Version 1.1 (released)

* Maintenance/fixes

## Version 1.0 (released)

* Initial public release.

src/J_Switchboard1_UI7.js
//# sourceURL=J_Switchboard_UI7.js
/**
 * J_Switchboard1_UI7.js
 * Configuration interface for Switchboard
 *
 * Copyright 2019,2020 Patrick H. Rigney, All Rights Reserved.
 * This file is part of Switchboard. For license information, see LICENSE at https://github.com/toggledbits/Switchboard
 */
/* globals api,jQuery,$,MultiBox */
/* jshint multistr: true */

//"use strict"; // fails on UI7, works fine with ALTUI

var Switchboard1_UI7 = (function(api, $) {

	var pluginVersion = "1.9-22203";

	var _UIVERSION = 21104; /* must agree with L_Switchboard1.lua */

	/* unique identifier for this plugin... */
	var uuid = 'fabe8224-2341-11e9-8762-74d4351650de'; /* 2019-01-28 Switchboard */

	var myModule = {};

	var serviceId = "urn:toggledbits-com:serviceId:Switchboard1";
	// var deviceType = "urn:schemas-toggledbits-com:device:Switchboard:1";

	var inStatusPanel = false;
	var isOpenLuup = false;
	var isALTUI = ( "undefined" !== typeof(MultiBox) );
	var childBehavior = {};

	/* Closing the control panel. */
	function onBeforeCpanelClose(args) {
		inStatusPanel = false;
	}

	/* Return footer */
	function footer() {
		var html = '';
		html += '<div class="clearfix">';
		html += '<div id="tbbegging"><em>Find Switchboard useful?</em> Please consider a small one-time donation to support this and my other plugins on <a href="https://www.toggledbits.com/donate" target="_blank">my web site</a>. I am grateful for any support you choose to give!</div>';
		html += '<div id="tbcopyright">Switchboard ver ' + pluginVersion + ' &copy; 2019 <a href="https://www.toggledbits.com/" target="_blank">Patrick H. Rigney</a>,' +
			' All Rights Reserved.' +
			' <a href="https://community.getvera.com/t/new-plugin-switchboard-virtual-switches-re-imagined/200515/1" target="_blank">Support Forum Thread</a> &#149; <a href="https://github.com/toggledbits/Switchboard-Vera/" target="_blank">Documentation and license information</a>.';
		try {
			html += '<div id="browserident">' + navigator.userAgent + '</div>';
		} catch( e ) {}

		return html;
	}

	function initModule() {
		var myid = api.getCpanelDeviceId();

		/* Check agreement of plugin core and UI */
		var s = api.getDeviceState( myid, serviceId, "_UIV", { dynamic: false } ) || "0";
		console.log("initModule() for device " + myid + " requires UI version " + _UIVERSION + ", seeing " + s);
		if ( String(_UIVERSION) != s ) {
			api.setCpanelContent( '<div style="border: 4px solid red; padding: 8px;">' +
				" ERROR! The Switchboard plugin core version and UI version do not agree." +
				" This may cause errors or corrupt your Switchboard configuration." +
				" Please hard-reload your browser and try again " +
				' (<a href="https://duckduckgo.com/?q=hard+reload+browser" target="_blank">how?</a>).' +
				" If you have installed hotfix patches, you may not have successfully installed all required files." +
				" Expected " + String(_UIVERSION) + " got " + String(s) +
				".</div>" );
			return false;
		}

		var dl = api.getListOfDevices();
		for ( var ix=0; ix<dl.length; ++ix ) {
			if ( dl[ix].device_type == "openLuup" && dl[ix].id_parent == 0 ) {
				isOpenLuup = true;
				break;
			}
		}

		api.registerEventHandler('on_ui_cpanel_before_close', Switchboard1_UI7, 'onBeforeCpanelClose');
		inStatusPanel = false;

		/* Load material design icons */
		jQuery("head").append('<link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">');
		jQuery("head").append('\
<style> \
	div#tbcopyright { display: block; margin: 12px 0px; } \
	div#tbbegging { display: block; color: #ff6600; margin-top: 12px; } \
</style>');

		return true;
	}

	function getDevices(pdev) {
		var dl = api.getListOfDevices();
		var dd = [];
		for ( var ix=0; ix<dl.length; ++ix ) {
			var dobj = dl[ix];
			if ( dobj.id_parent == pdev ) {
				dd.push( dobj );
			}
		}
		dd.sort( function( a, b ) {
			var an = a.name.toLowerCase();
			var bn = b.name.toLowerCase();
			if ( an == bn ) return 0;
			return an < bn ? -1 : 1;
		});
		return dd;
	}

	function handleStateClick( ev ) {
		var el = jQuery( ev.currentTarget );
		var row = el.closest( 'div.row' );
		var dev = parseInt( row.attr( 'id' ).replace( /^d/i, "" ) );
		api.performActionOnDevice( dev, "urn:micasaverde-com:serviceId:HaDevice1", "ToggleState", { actionArguments: {} });
	}

	function handleIconClick( ev ) {
		var el = jQuery( ev.currentTarget );
		var row = el.closest( 'div.row' );
		var dev = parseInt( row.attr( 'id' ).replace( /^d/i, "" ) );
		var act = el.data( 'swbd-action' );
		var t;

		switch ( act ) {
			case 'visibility':
				var vis = el.text() == "visibility";
				/* Changing the device property this way causes a Luup reload. So don't do that. We've got our own way. */
				/* api.setDeviceProperty( dev, 'invisible', vis ? 1 : 0, { persistent: true } ); */
				var devobj = api.getDeviceObject( dev );
				api.performActionOnDevice( devobj.id_parent, serviceId, "SetSwitchVisibility", { actionArguments: { DeviceNum: dev, Visibility: vis ? "0" : "1" } });
				el.text( vis ? "visibility_off" : "visibility" );
				break;

			case 'impulse':
				t = api.getDeviceState( dev, serviceId, "ImpulseTime" ) || "0";
				while (true) {
					t = prompt( 'Please enter pulse/reset time (secs, 0=normal operation):', t );
					if ( t == null ) break;
					var n = parseInt( t );
					if ( ! isNaN( n ) && n >= 0 ) {
						api.setDeviceStatePersistent( dev, serviceId, "ImpulseTime", String(n) );
						el.text( n == 0 ? "timer_off" : "timer" );
						break;
					}
				}
				break;

			case 'repeat':
				t = 0 !== parseInt( api.getDeviceState( dev, serviceId, "AlwaysUpdateStatus" ) || 0 );
				/* Set opposite icon and value */
				jQuery( 'i.swbd-repeat', row ).text( t ? 'repeat_one' : 'repeat' )
					.attr( 'title', t ? "Trigger only if status changes" : "Always trigger" );
				api.setDeviceStatePersistent( dev, serviceId, "AlwaysUpdateStatus", String(t ? 0 : 1) );
				break;

			default:
				console.log("handleIconClick() unhandled action " + String(act));
		}
	}

	function handleTextClick( ev ) {
		var el = jQuery( ev.currentTarget );
		var row = el.closest( 'div.row' );
		var dev = parseInt( row.attr( 'id' ).replace( /^d/i, "" ) );
		var subject = el.hasClass( 'vstext2' ) ? 2 : 1;
		var txt = api.getDeviceState( dev, "urn:upnp-org:serviceId:VSwitch1", subject ) || "";
		var newText = prompt( 'Enter new Text'+subject+' value:', txt );
		console.log( typeof(newText) + ":" + String(newText) );
		if ( newText != null ) {
			api.setDeviceStatePersistent( dev, "urn:upnp-org:serviceId:VSwitch1", 'Text'+subject, newText );
			jQuery( 'span.vstext' + subject, row ).text( newText );
		}
	}

	function handleNameClick( ev ) {
		var el = jQuery( ev.currentTarget );
		var row = el.closest( 'div.row' );
		var dev = parseInt( row.attr( 'id' ).replace( /^d/i, "" ) );
		var devobj = api.getDeviceObject( dev );
		if ( devobj ) {
			var txt = devobj.name;
			while ( true ) {
				txt = prompt( 'Enter new switch name:', txt );
				if ( txt == null ) break;
				if ( txt.match( /^.+$/ ) ) {
					/* This causes a Luup reload, so don't do it this way. We have our own way. */
					/* api.setDeviceProperty( dev, 'name', txt, { persistent: true } ); */
					api.performActionOnDevice( devobj.id_parent, serviceId, "SetSwitchName", { actionArguments: { DeviceNum: dev, NewName: txt } });
					el.text( txt );
					break;
				}
			}
		}
	}

	function updateStatus(pdev) {
		pdev = pdev || api.getCpanelDeviceId();
		var container = jQuery( "div#switchboardstatus div#devices" ).empty();

		var switches = getDevices( pdev );
		var row = jQuery('<div class="row headrow"></div>');
		row.append( '<div class="colhead col-xs-1 col-md-1"></div>' );
		row.append( '<div class="colhead col-xs-11 col-md-5">Device Name</div>' );
		row.append( '<div class="colhead col-xs-4 col-md-2">Options</div>' );
		row.append( '<div class="colhead col-xs-4 col-md-2">Text1</div>' );
		row.append( '<div class="colhead col-xs-4 col-md-2">Text2</div>' );
		container.append( row );

		jQuery.each( switches, function( ix, obj ) {
			var behavior = api.getDeviceState( obj.id, serviceId, "Behavior" ) || "Binary";
			row = jQuery('<div class="row devicerow"></div>');
			var el = jQuery( '<div class="col-xs-1 col-md-1 text-right"></div>' );
			var st = api.getDeviceState( obj.id, "urn:upnp-org:serviceId:SwitchPower1", "Status" ) || "0";
			if ( "Cover" === behavior ) {
				var l = parseInt( api.getDeviceState( obj.id, "urn:upnp-org:serviceId:Dimming1", "LoadLevelStatus" ) || "0" );
				l = Math.floor( l / 10 + 0.5 ) * 10;
				el.append( '<img class="swbd-state" src="/cmh/skins/default/img/devices/device_states/window_covering_' +
					( st == "0" ? "off" : l ) + '.png" width="32" height="32" alt="cover state">' );
			} else {
				el.append( '<img class="swbd-state" src="https://www.toggledbits.com/assets/switchboard/switchboard-switch-' +
					( {"0":"off","1":"on","2":"x"}[st] || "x" ) + '.png" width="32" height="32" alt="switch state">' );
			}
			el.attr( 'title', 'Click to toggle state' );
			row.append( el );
			row.append( jQuery( '<div class="vsname col-xs-11 col-md-5"></div>' ).text( obj.name + ' (#' + obj.id + ')' ).attr( 'title', 'Click to change name' ) );
			el = jQuery( '<div class="col-xs-4 col-md-2"></div>' );
			jQuery( '<i class="swbd-vis material-icons md-btn" title="Toggle visibility">visibility</i>' )
				.data( 'swbd-action', 'visibility' )
				.appendTo( el );
			if ( false !== ( childBehavior[behavior] || {} ).timer ) {
				jQuery( '<i class="swbd-impulse material-icons md-btn" title="Set auto-reset timer">timer_off</i>' )
					.data('swbd-action', 'impulse')
					.appendTo( el );
			}
			if ( isOpenLuup ) {
				jQuery( '<i class="swbd-repeat material-icons md-btn" title="Trigger only if status changes">repeat_one</i>' )
					.data('swbd-action', 'repeat')
					.appendTo( el );
			}
			row.append( el );

			if ( "Binary" === behavior ) {
				var s = api.getDeviceState( obj.id, "urn:upnp-org:serviceId:VSwitch1", "Text1" );
				row.append( '<div class="col-xs-4 col-md-2"><span class="vstext vstext1"></span><i id="vstext1" class="vstext material-icons md-btn">create</i></div>' );
				jQuery( 'span.vstext1', row ).text( s || "" );
				s = api.getDeviceState( obj.id, "urn:upnp-org:serviceId:VSwitch1", "Text2" );
				row.append( '<div class="col-xs-4 col-md-2"><span class="vstext vstext2"></span><i id="vstext2" class="vstext material-icons md-btn">create</i></div>' );
				jQuery( 'span.vstext2', row ).text( s || "" );
			} else {
				row.append( '<div class="col-xs-8 col-md-4"></div>' );
			}

			row.attr( 'id', 'd' + obj.id );
			container.append( row );

			if ( ( obj.invisible || "0" ) != "0" ) {
				jQuery( 'i.swbd-vis', row ).text( 'visibility_off' );
			}
			st = parseInt( api.getDeviceState( obj.id, serviceId, "ImpulseTime" ) || 0 );
			if ( ! isNaN( st ) && st > 0 ) {
				jQuery( 'i.swbd-impulse', row ).text( 'timer' )
					.attr( 'title', 'Auto-reset ' + st + ' seconds; click to modify or clear' );
			}
			st = 0 !== parseInt( api.getDeviceState( obj.id, serviceId, "AlwaysUpdateStatus" ) || 0 );
			if ( st ) {
				jQuery( 'i.swbd-repeat', row ).text( 'repeat' );
			}

			jQuery( 'img.swbd-state', row ).on( 'click.switchboard', handleStateClick );
			jQuery( 'i.md-btn', row ).on( 'click.switchboard', handleIconClick );
			jQuery( '.vstext', row ).attr( 'title', 'Click to edit' ).on( 'click.switchboard', handleTextClick );
			jQuery( 'div.vsname', row ).on( 'click.switchboard', handleNameClick );
		});
	}

	function onUIDeviceStatusChanged( args ) {
		if ( !inStatusPanel ) {
			return;
		}
		var pdev = api.getCpanelDeviceId();
		var ix = api.getDeviceObject( args.id );
		if ( ix && ix.id_parent == pdev ) {
			updateStatus( pdev );
		}
	}

	function waitForReload() {
		jQuery.ajax({
			url: api.getDataRequestURL(),
			data: {
				id: "status",
				DeviceNum: api.getCpanelDeviceId(),
				output_format: "json"
			},
			dataType: "json",
			timeout: 5000
		}).done( function( data, statusText, jqXHR ) {
			var key = "Device_Num_" + api.getCpanelDeviceId();
			if ( data[key] && -1 === parseInt( data[key].status ) ) {
				jQuery( 'div#tail button#addchild' ).prop( 'disabled', false );
				jQuery( 'div#tail div#notice' ).text("");
				setTimeout( updateStatus, 2000 );
			} else {
				jQuery( 'div#tail div#notice' ).append( "&ndash;" );
				setTimeout( waitForReload, 1000 );
			}
		}).fail( function( jqXHR, textStatus, errorThrown ) {
			jQuery( 'div#tail div#notice' ).append( "&bull;" );
			setTimeout( waitForReload, 2000 );
		});
	}

	function handleAddChildClick( ev ) {
		var el = jQuery( ev.currentTarget );
		var row = el.closest( 'div#tail' );
		var childType = jQuery( 'select#childtype', row ).val() || "";
		if ( "" !== childType ) {
			api.performActionOnDevice( api.getCpanelDeviceId(), serviceId, "AddChild", {
				actionArguments: { DeviceType: childType },
				onSuccess: function( xhr ) {
					el.prop( 'disabled', true );
					jQuery( 'div#notice', row ).text("Creating child... please wait while Luup reloads...");
					setTimeout( waitForReload, 5000 );
				},
				onFailure: function( xhr ) {
					alert( "An error occurred. Try again in a moment; Vera may be busy." );
				}
			} );
		} else {
			jQuery( 'div#notice', row ).text("Wählen Sie zuerst einen Typ!");
		}
	}

	function doStatusPanel()
	{
		console.log("doStatusPanel()");

		try {
			if ( ! initModule() ) {
				return;
			}

			/* Our styles. */
			var html = "<style>";
			html += 'div#switchboardstatus {}';
			html += 'div#switchboardstatus div.row { min-height: 40px; margin-top: 4px; margin-bottom: 4px; border-bottom: 1px dotted #006040; }';
			html += 'div#switchboardstatus div.vsname { font-size: 16px; }';
			html += 'div#switchboardstatus i.md-btn { margin-right: 4px; }';
			html += 'div#switchboardstatus i.vstext.md-btn { font-size: 18px; }';
			html += 'div#switchboardstatus div.headrow { color: white; font-size: 16px; font-weight: bold; line-height: 40px; background-color: #00a652; }';
			html += 'div#switchboardstatus div.colhead { }';
			html += "</style>";
			jQuery("head").append( html );

			html = '<div id="switchboardstatus" class="switchboardtab"></div>';
			html += footer();
			api.setCpanelContent( html );

			var container = jQuery( "div#switchboardstatus" );
			container.append( jQuery( '<div id="devices">Loading... please wait...</div>' ) );
			var br = jQuery( '<div id="tail" class="form-inline"></div>' );
			var sel = jQuery( '<select id="childtype" class="form-control form-control-sm"></select>' );
			sel.append( jQuery( '<option></option>' ).val("").text('--choose type--') ).val( "" ); /* default */
			br.append( sel );
			br.append( '<button id="addchild" class="btn btn-md btn-primary">Create New Virtual Device</button>' );
			br.append( '<div id="notice"></div>' );
			container.append( br );

			/* Now, populate the menu */
			jQuery( 'button#addchild', container ).on( 'click.switchboard', handleAddChildClick ).prop( 'disabled', true );
			jQuery.ajax({
				url: api.getDataRequestURL(),
				data: {
					id: "lr_Switchboard",
					action: "getvtypes"
				},
				dataType: "json",
				timeout: 5000
			}).done( function( data, statusText, jqXHR ) {
				childBehavior = data;
				var hasOne = false;
				var childMenu = jQuery( 'div#tail select#childtype' );
				var lx = [];
				var ch;
				for ( ch in data ) {
					if ( data.hasOwnProperty( ch ) ) {
						lx[lx.length] = ch;
					}
				}
				lx.sort( function( a, b ) {
					var p1 = data[a].order || 32767;
					var p2 = data[b].order || 32767;
					if ( p1 == p2 ) {
						/* the names will never be equal */
						return data[a].name < data[b].name ? -1 : 1;
					}
					return p1 - p2;
				});
				for ( var k=0; k<lx.length; k++) {
					ch = lx[k];
					childMenu.append( jQuery( '<option></option>' ).val( ch ).text( data[ch].name || ch ) );
					hasOne = true;
				}
				if ( lx.length > 0 ) {
					jQuery( 'div#tail button#addchild' ).prop( 'disabled', false );
				}

				updateStatus( api.getCpanelDeviceId() );

				api.registerEventHandler('on_ui_deviceStatusChanged', Switchboard1_UI7, 'onUIDeviceStatusChanged');
				inStatusPanel = true; /* Tell the event handler it's OK */

			}).fail( function( jqXHR ) {
				alert( "There was an error loading configuration data. Vera may be busy; try again in a moment." );
			});
		}
		catch( e ) {
			alert( String(e) + "\n" + e.stack );
		}
	}


/** ***************************************************************************
 *
 * C L O S I N G
 *
 ** **************************************************************************/

	console.log("Initializing Switchboard1_UI7 module");

	myModule = {
		uuid: uuid,
		initModule: initModule,
		onBeforeCpanelClose: onBeforeCpanelClose,
		onUIDeviceStatusChanged: onUIDeviceStatusChanged,
		doStatusPanel: doStatusPanel
	};
	return myModule;
})(api, $ || jQuery);
